<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>CodonU.analyzer.internal_comp &mdash; CodonU 1.1.0 documentation</title>
      <link rel="stylesheet" href="../../../../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../../../../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../../../../_static/graphviz.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../../../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../../../../" id="documentation_options" src="../../../../_static/documentation_options.js"></script>
        <script src="../../../../_static/doctools.js"></script>
        <script src="../../../../_static/sphinx_highlight.js"></script>
    <script src="../../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../../genindex.html" />
    <link rel="search" title="Search" href="../../../../search.html" />
    <link rel="next" title="CodonU.analyzer.rscu_comp" href="../rscu_comp/index.html" />
    <link rel="prev" title="CodonU.analyzer.gravy_comp" href="../gravy_comp/index.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../../index.html" class="icon icon-home">
            CodonU
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="../../../index.html">API Reference</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="../../index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">CodonU</span></code></a><ul class="current">
<li class="toctree-l3 current"><a class="reference internal" href="../../index.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l4 current"><a class="reference internal" href="../index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">CodonU.analyzer</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../correspondence_analysis/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">CodonU.correspondence_analysis</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../cua_errors/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">CodonU.cua_errors</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../cua_warnings/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">CodonU.cua_warnings</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../extractor/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">CodonU.extractor</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../file_handler/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">CodonU.file_handler</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../phylogenetic_analysis/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">CodonU.phylogenetic_analysis</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../vizualizer/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">CodonU.vizualizer</span></code></a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../../index.html">CodonU</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../../index.html">API Reference</a></li>
          <li class="breadcrumb-item"><a href="../../index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">CodonU</span></code></a></li>
          <li class="breadcrumb-item"><a href="../index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">CodonU.analyzer</span></code></a></li>
      <li class="breadcrumb-item active"><code class="xref py py-mod docutils literal notranslate"><span class="pre">CodonU.analyzer.internal_comp</span></code></li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../../../_sources/autoapi/CodonU/analyzer/internal_comp/index.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="module-CodonU.analyzer.internal_comp">
<span id="codonu-analyzer-internal-comp"></span><h1><a class="reference internal" href="#module-CodonU.analyzer.internal_comp" title="CodonU.analyzer.internal_comp"><code class="xref py py-mod docutils literal notranslate"><span class="pre">CodonU.analyzer.internal_comp</span></code></a><a class="headerlink" href="#module-CodonU.analyzer.internal_comp" title="Permalink to this heading"></a></h1>
<section id="module-contents">
<h2>Module Contents<a class="headerlink" href="#module-contents" title="Permalink to this heading"></a></h2>
<section id="functions">
<h3>Functions<a class="headerlink" href="#functions" title="Permalink to this heading"></a></h3>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#CodonU.analyzer.internal_comp.is_not_bad_seq" title="CodonU.analyzer.internal_comp.is_not_bad_seq"><code class="xref py py-obj docutils literal notranslate"><span class="pre">is_not_bad_seq</span></code></a>(→ bool)</p></td>
<td><p>Checks if the sequence is bad i.e. length of the sequence is not divisible by 3</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#CodonU.analyzer.internal_comp.not_contains_amb_letter" title="CodonU.analyzer.internal_comp.not_contains_amb_letter"><code class="xref py py-obj docutils literal notranslate"><span class="pre">not_contains_amb_letter</span></code></a>(→ bool)</p></td>
<td><p>Checks if provided sequence contains ambiguous DNA letters</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#CodonU.analyzer.internal_comp.g3" title="CodonU.analyzer.internal_comp.g3"><code class="xref py py-obj docutils literal notranslate"><span class="pre">g3</span></code></a>(→ float)</p></td>
<td><p>Calculates percentage of G content for third position</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#CodonU.analyzer.internal_comp.a3" title="CodonU.analyzer.internal_comp.a3"><code class="xref py py-obj docutils literal notranslate"><span class="pre">a3</span></code></a>(→ float)</p></td>
<td><p>Calculates percentage of A content for third position</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#CodonU.analyzer.internal_comp.gc_123" title="CodonU.analyzer.internal_comp.gc_123"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gc_123</span></code></a>(→ tuple[float, float | int, float | int, ...)</p></td>
<td><p>Calculate G+C content: total, for first, second and third positions</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#CodonU.analyzer.internal_comp.at_123" title="CodonU.analyzer.internal_comp.at_123"><code class="xref py py-obj docutils literal notranslate"><span class="pre">at_123</span></code></a>(→ tuple[float, float | int, float | int, ...)</p></td>
<td><p>Calculate G+C content: total, for first, second and third positions</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#CodonU.analyzer.internal_comp.custom_codon_table" title="CodonU.analyzer.internal_comp.custom_codon_table"><code class="xref py py-obj docutils literal notranslate"><span class="pre">custom_codon_table</span></code></a>(→ None)</p></td>
<td><p>Registers a new Codon Table as provided by the user.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#CodonU.analyzer.internal_comp.filter_reference" title="CodonU.analyzer.internal_comp.filter_reference"><code class="xref py py-obj docutils literal notranslate"><span class="pre">filter_reference</span></code></a>(→ list[Bio.SeqRecord.SeqRecord])</p></td>
<td><p>Filters the list of reference based on given threshold of length</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#CodonU.analyzer.internal_comp.reverse_table" title="CodonU.analyzer.internal_comp.reverse_table"><code class="xref py py-obj docutils literal notranslate"><span class="pre">reverse_table</span></code></a>(→ dict[str, list[str]])</p></td>
<td><p>Creates the protein, codon dictionary where protein is key</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#CodonU.analyzer.internal_comp.syn_codons" title="CodonU.analyzer.internal_comp.syn_codons"><code class="xref py py-obj docutils literal notranslate"><span class="pre">syn_codons</span></code></a>(→ dict[str, list[str]])</p></td>
<td><p>Creates the codon, synonymous codon family dictionary where codon is the key</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#CodonU.analyzer.internal_comp.sf_vals" title="CodonU.analyzer.internal_comp.sf_vals"><code class="xref py py-obj docutils literal notranslate"><span class="pre">sf_vals</span></code></a>(→ dict[int, list[str]])</p></td>
<td><p>Creates the sf value and protein dictionary where sf value is key</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#CodonU.analyzer.internal_comp.rscu" title="CodonU.analyzer.internal_comp.rscu"><code class="xref py py-obj docutils literal notranslate"><span class="pre">rscu</span></code></a>(→ dict[str, float])</p></td>
<td><p>Calculates relative synonymous codon usage (RSCU) value for a given nucleotide sequence according to Sharp and Li (1987)</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#CodonU.analyzer.internal_comp.weights_for_cai" title="CodonU.analyzer.internal_comp.weights_for_cai"><code class="xref py py-obj docutils literal notranslate"><span class="pre">weights_for_cai</span></code></a>(→ dict[str, float])</p></td>
<td><p>Calculates relative adaptiveness/weight value for a given nucleotide sequence according to Sharp and Li (1987)</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#CodonU.analyzer.internal_comp.cai" title="CodonU.analyzer.internal_comp.cai"><code class="xref py py-obj docutils literal notranslate"><span class="pre">cai</span></code></a>(→ float)</p></td>
<td><p>Calculates Codon Adaptive Index (CAI) value for a given nucleotide sequence according to Sharp and Li (1987)</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#CodonU.analyzer.internal_comp.cbi" title="CodonU.analyzer.internal_comp.cbi"><code class="xref py py-obj docutils literal notranslate"><span class="pre">cbi</span></code></a>(→ tuple[float, str])</p></td>
<td><p>Calculates codon bias index (CBI) for a given protein seq based on Bennetzen and Hall (1982)</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#CodonU.analyzer.internal_comp.enc" title="CodonU.analyzer.internal_comp.enc"><code class="xref py py-obj docutils literal notranslate"><span class="pre">enc</span></code></a>(→ float)</p></td>
<td><p>Calculates Effective number of codons (Enc) based on Wright (1989) and Fuglsang (2004)</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#CodonU.analyzer.internal_comp.gravy" title="CodonU.analyzer.internal_comp.gravy"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gravy</span></code></a>(→ float)</p></td>
<td><p>Computes the GRAVY score according to Kyte and Doolittle (1982)</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#CodonU.analyzer.internal_comp.aromaticity" title="CodonU.analyzer.internal_comp.aromaticity"><code class="xref py py-obj docutils literal notranslate"><span class="pre">aromaticity</span></code></a>(→ float)</p></td>
<td><p>Calculate the aromaticity score according to Lobry (1994).</p></td>
</tr>
</tbody>
</table>
<dl class="py function">
<dt class="sig sig-object py" id="CodonU.analyzer.internal_comp.is_not_bad_seq">
<span class="sig-prename descclassname"><span class="pre">CodonU.analyzer.internal_comp.</span></span><span class="sig-name descname"><span class="pre">is_not_bad_seq</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Seq</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">code</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">_type</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">bool</span></span></span><a class="headerlink" href="#CodonU.analyzer.internal_comp.is_not_bad_seq" title="Permalink to this definition"></a></dt>
<dd><p>Checks if the sequence is bad i.e. length of the sequence is not divisible by 3</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seq</strong> – The nucleotide sequence</p></li>
<li><p><strong>code</strong> – The code to call BadSequenceError (1 or 2)</p></li>
<li><p><strong>_type</strong> – Type of sequence, i.e. ‘nuc’</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>True if seq is not bad</p>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="../index.html#CodonU.analyzer.BadSequenceError" title="CodonU.analyzer.BadSequenceError"><strong>BadSequenceError</strong></a> – If the seq is bad</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="CodonU.analyzer.internal_comp.not_contains_amb_letter">
<span class="sig-prename descclassname"><span class="pre">CodonU.analyzer.internal_comp.</span></span><span class="sig-name descname"><span class="pre">not_contains_amb_letter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Seq</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">bool</span></span></span><a class="headerlink" href="#CodonU.analyzer.internal_comp.not_contains_amb_letter" title="Permalink to this definition"></a></dt>
<dd><p>Checks if provided sequence contains ambiguous DNA letters</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>seq</strong> – Provided sequence</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>True if sequence does not contain ambiguous letter</p>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="../index.html#CodonU.analyzer.NucleotideError" title="CodonU.analyzer.NucleotideError"><strong>NucleotideError</strong></a> – If sequence contain ambiguous letter</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="CodonU.analyzer.internal_comp.g3">
<span class="sig-prename descclassname"><span class="pre">CodonU.analyzer.internal_comp.</span></span><span class="sig-name descname"><span class="pre">g3</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Seq</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">float</span></span></span><a class="headerlink" href="#CodonU.analyzer.internal_comp.g3" title="Permalink to this definition"></a></dt>
<dd><p>Calculates percentage of G content for third position</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>seq</strong> – Provided sequence</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Percentage of G content</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="CodonU.analyzer.internal_comp.a3">
<span class="sig-prename descclassname"><span class="pre">CodonU.analyzer.internal_comp.</span></span><span class="sig-name descname"><span class="pre">a3</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Seq</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">float</span></span></span><a class="headerlink" href="#CodonU.analyzer.internal_comp.a3" title="Permalink to this definition"></a></dt>
<dd><p>Calculates percentage of A content for third position</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>seq</strong> – Provided sequence</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Percentage of A content</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="CodonU.analyzer.internal_comp.gc_123">
<span class="sig-prename descclassname"><span class="pre">CodonU.analyzer.internal_comp.</span></span><span class="sig-name descname"><span class="pre">gc_123</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Seq</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">float</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">float</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">int</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">float</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">int</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">float</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">int</span><span class="p"><span class="pre">]</span></span></span></span><a class="headerlink" href="#CodonU.analyzer.internal_comp.gc_123" title="Permalink to this definition"></a></dt>
<dd><p>Calculate G+C content: total, for first, second and third positions</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>seq</strong> – Provided sequence</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The G+C percentage for the entire sequence, and the three codon positions</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="CodonU.analyzer.internal_comp.at_123">
<span class="sig-prename descclassname"><span class="pre">CodonU.analyzer.internal_comp.</span></span><span class="sig-name descname"><span class="pre">at_123</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Seq</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">float</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">float</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">int</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">float</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">int</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">float</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">int</span><span class="p"><span class="pre">]</span></span></span></span><a class="headerlink" href="#CodonU.analyzer.internal_comp.at_123" title="Permalink to this definition"></a></dt>
<dd><p>Calculate G+C content: total, for first, second and third positions</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>seq</strong> – Provided sequence</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The A+T percentage for the entire sequence, and the three codon positions</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="CodonU.analyzer.internal_comp.custom_codon_table">
<span class="sig-prename descclassname"><span class="pre">CodonU.analyzer.internal_comp.</span></span><span class="sig-name descname"><span class="pre">custom_codon_table</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alt_name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genetic_code_id</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">forward_table</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start_codons</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stop_codons</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">None</span></span></span><a class="headerlink" href="#CodonU.analyzer.internal_comp.custom_codon_table" title="Permalink to this definition"></a></dt>
<dd><p>Registers a new Codon Table as provided by the user.</p>
<p><strong>Note</strong>: The scope of the newly registered table is limited to the working file only</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>name</strong> – Name for the table</p></li>
<li><p><strong>alt_name</strong> – Short name for the table</p></li>
<li><p><strong>genetic_code_id</strong> – Genetic code number for the table</p></li>
<li><p><strong>forward_table</strong> – A dict containing mapping of codons to proteins [excluding stop codons]</p></li>
<li><p><strong>start_codons</strong> – A list of possible start codons</p></li>
<li><p><strong>stop_codons</strong> – A list of possible stop codons</p></li>
</ul>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="../index.html#CodonU.analyzer.CodonTableExistsError" title="CodonU.analyzer.CodonTableExistsError"><strong>CodonTableExistsError</strong></a> – If the name, alt_name or genetic_code_id already exists</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="CodonU.analyzer.internal_comp.filter_reference">
<span class="sig-prename descclassname"><span class="pre">CodonU.analyzer.internal_comp.</span></span><span class="sig-name descname"><span class="pre">filter_reference</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">records</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_len_threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">_type</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">Bio.SeqRecord.SeqRecord</span><span class="p"><span class="pre">]</span></span></span></span><a class="headerlink" href="#CodonU.analyzer.internal_comp.filter_reference" title="Permalink to this definition"></a></dt>
<dd><p>Filters the list of reference based on given threshold of length</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>records</strong> – A generator object holding the sequence objects</p></li>
<li><p><strong>min_len_threshold</strong> – Minimum length of nucleotide sequence to be considered as gene</p></li>
<li><p><strong>_type</strong> – Type of sequence, i.e. ‘nuc’ or ‘aa</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The list of usable sequences</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="CodonU.analyzer.internal_comp.reverse_table">
<span class="sig-prename descclassname"><span class="pre">CodonU.analyzer.internal_comp.</span></span><span class="sig-name descname"><span class="pre">reverse_table</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">codon_table</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Bio.Data.CodonTable.NCBICodonTableDNA</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="headerlink" href="#CodonU.analyzer.internal_comp.reverse_table" title="Permalink to this definition"></a></dt>
<dd><p>Creates the protein, codon dictionary where protein is key</p>
<p>e.g. ‘L’: [‘TTA’, ‘TTG’, ‘CTT’, ‘CTC’, ‘CTA’, ‘CTG’]</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>codon_table</strong> – The codon table</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The dict having protein as key</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="CodonU.analyzer.internal_comp.syn_codons">
<span class="sig-prename descclassname"><span class="pre">CodonU.analyzer.internal_comp.</span></span><span class="sig-name descname"><span class="pre">syn_codons</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">codon_table</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Bio.Data.CodonTable.NCBICodonTableDNA</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="headerlink" href="#CodonU.analyzer.internal_comp.syn_codons" title="Permalink to this definition"></a></dt>
<dd><p>Creates the codon, synonymous codon family dictionary where codon is the key</p>
<p>e.g. ‘TTA’: [‘TTA’, ‘TTG’, ‘CTT’, ‘CTC’, ‘CTA’, ‘CTG’]</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>codon_table</strong> – The codon table</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The dict having individual codons as keys</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="CodonU.analyzer.internal_comp.sf_vals">
<span class="sig-prename descclassname"><span class="pre">CodonU.analyzer.internal_comp.</span></span><span class="sig-name descname"><span class="pre">sf_vals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">codon_table</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Bio.Data.CodonTable.NCBICodonTableDNA</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="headerlink" href="#CodonU.analyzer.internal_comp.sf_vals" title="Permalink to this definition"></a></dt>
<dd><p>Creates the sf value and protein dictionary where sf value is key</p>
<p>e.g. 6: [‘L’, ‘S’, ‘R’]</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>codon_table</strong> – The codon table</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The dict having sf values as key</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="CodonU.analyzer.internal_comp.rscu">
<span class="sig-prename descclassname"><span class="pre">CodonU.analyzer.internal_comp.</span></span><span class="sig-name descname"><span class="pre">rscu</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">references</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">Seq</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genetic_code</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">float</span><span class="p"><span class="pre">]</span></span></span></span><a class="headerlink" href="#CodonU.analyzer.internal_comp.rscu" title="Permalink to this definition"></a></dt>
<dd><p>Calculates relative synonymous codon usage (RSCU) value for a given nucleotide sequence according to Sharp and Li (1987)</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>references</strong> – List of reference nucleotide sequences</p></li>
<li><p><strong>genetic_code</strong> – Genetic table number for codon table</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A dictionary containing codons and their respective RSCU values</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="CodonU.analyzer.internal_comp.weights_for_cai">
<span class="sig-prename descclassname"><span class="pre">CodonU.analyzer.internal_comp.</span></span><span class="sig-name descname"><span class="pre">weights_for_cai</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">references</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">Seq</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genetic_code</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">float</span><span class="p"><span class="pre">]</span></span></span></span><a class="headerlink" href="#CodonU.analyzer.internal_comp.weights_for_cai" title="Permalink to this definition"></a></dt>
<dd><p>Calculates relative adaptiveness/weight value for a given nucleotide sequence according to Sharp and Li (1987)</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>references</strong> – List of reference nucleotide sequences</p></li>
<li><p><strong>genetic_code</strong> – Genetic table number for codon table</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A dictionary containing codons and their respective weights</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="CodonU.analyzer.internal_comp.cai">
<span class="sig-prename descclassname"><span class="pre">CodonU.analyzer.internal_comp.</span></span><span class="sig-name descname"><span class="pre">cai</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">nuc_seq</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Seq</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">references</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">Seq</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genetic_code</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">float</span></span></span><a class="headerlink" href="#CodonU.analyzer.internal_comp.cai" title="Permalink to this definition"></a></dt>
<dd><p>Calculates Codon Adaptive Index (CAI) value for a given nucleotide sequence according to Sharp and Li (1987)</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>nuc_seq</strong> – The Nucleotide Sequence</p></li>
<li><p><strong>references</strong> – List of reference nucleotide sequences</p></li>
<li><p><strong>genetic_code</strong> – Genetic table number for codon table</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The CAI value for given sequence</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="CodonU.analyzer.internal_comp.cbi">
<span class="sig-prename descclassname"><span class="pre">CodonU.analyzer.internal_comp.</span></span><span class="sig-name descname"><span class="pre">cbi</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">prot_seq</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Seq</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">references</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">Seq</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genetic_code</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">float</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></span><a class="headerlink" href="#CodonU.analyzer.internal_comp.cbi" title="Permalink to this definition"></a></dt>
<dd><p>Calculates codon bias index (CBI) for a given protein seq based on Bennetzen and Hall (1982)</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>prot_seq</strong> – The Protein Sequence</p></li>
<li><p><strong>references</strong> – List of reference nucleotide sequences</p></li>
<li><p><strong>genetic_code</strong> – Genetic table number for codon table</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A tuple of CBI val and the optimal codon</p>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><a class="reference internal" href="../index.html#CodonU.analyzer.NoSynonymousCodonWarning" title="CodonU.analyzer.NoSynonymousCodonWarning"><strong>NoSynonymousCodonWarning</strong></a> – When there is no synonymous codons</p></li>
<li><p><a class="reference internal" href="../index.html#CodonU.analyzer.MissingCodonWarning" title="CodonU.analyzer.MissingCodonWarning"><strong>MissingCodonWarning</strong></a> – When no codons translate to provided Amino acid</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="CodonU.analyzer.internal_comp.enc">
<span class="sig-prename descclassname"><span class="pre">CodonU.analyzer.internal_comp.</span></span><span class="sig-name descname"><span class="pre">enc</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">references</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">Seq</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genetic_code</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">float</span></span></span><a class="headerlink" href="#CodonU.analyzer.internal_comp.enc" title="Permalink to this definition"></a></dt>
<dd><p>Calculates Effective number of codons (Enc) based on Wright (1989) and Fuglsang (2004)</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>references</strong> – List of reference nucleotide sequences</p></li>
<li><p><strong>genetic_code</strong> – Genetic table number for codon table</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Calculated Enc value for the sequence(s)</p>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><a class="reference internal" href="../index.html#CodonU.analyzer.MissingCodonWarning" title="CodonU.analyzer.MissingCodonWarning"><strong>MissingCodonWarning</strong></a> – If there is no codon for a certain amino acid</p></li>
<li><p><a class="reference internal" href="../index.html#CodonU.analyzer.NoProteinError" title="CodonU.analyzer.NoProteinError"><strong>NoProteinError</strong></a> – If there is no codon for a certain set of amino acid</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="CodonU.analyzer.internal_comp.gravy">
<span class="sig-prename descclassname"><span class="pre">CodonU.analyzer.internal_comp.</span></span><span class="sig-name descname"><span class="pre">gravy</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Seq</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">float</span></span></span><a class="headerlink" href="#CodonU.analyzer.internal_comp.gravy" title="Permalink to this definition"></a></dt>
<dd><p>Computes the GRAVY score according to Kyte and Doolittle (1982)</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>seq</strong> – Protein sequence</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The GRAVY score</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="CodonU.analyzer.internal_comp.aromaticity">
<span class="sig-prename descclassname"><span class="pre">CodonU.analyzer.internal_comp.</span></span><span class="sig-name descname"><span class="pre">aromaticity</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Seq</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">float</span></span></span><a class="headerlink" href="#CodonU.analyzer.internal_comp.aromaticity" title="Permalink to this definition"></a></dt>
<dd><p>Calculate the aromaticity score according to Lobry (1994).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>seq</strong> – Protein sequence</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The aromaticity score</p>
</dd>
</dl>
</dd></dl>

</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="../gravy_comp/index.html" class="btn btn-neutral float-left" title="CodonU.analyzer.gravy_comp" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="../rscu_comp/index.html" class="btn btn-neutral float-right" title="CodonU.analyzer.rscu_comp" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Souradipto Choudhuri.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>