<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>CodonU.analyzer.tRNA_comp &mdash; CodonU 1.1.0 documentation</title>
      <link rel="stylesheet" href="../../../../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../../../../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../../../../_static/graphviz.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../../../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../../../../_static/documentation_options.js?v=fc837d61"></script>
        <script src="../../../../_static/doctools.js?v=888ff710"></script>
        <script src="../../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../../genindex.html" />
    <link rel="search" title="Search" href="../../../../search.html" />
    <link rel="next" title="CodonU.correspondence_analysis" href="../../correspondence_analysis/index.html" />
    <link rel="prev" title="CodonU.analyzer.rscu_comp" href="../rscu_comp/index.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../../index.html" class="icon icon-home">
            CodonU
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="../../../index.html">API Reference</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="../../index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">CodonU</span></code></a><ul class="current">
<li class="toctree-l3 current"><a class="reference internal" href="../../index.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l4 current"><a class="reference internal" href="../index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">CodonU.analyzer</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../correspondence_analysis/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">CodonU.correspondence_analysis</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../cua_errors/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">CodonU.cua_errors</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../cua_warnings/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">CodonU.cua_warnings</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../extractor/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">CodonU.extractor</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../file_handler/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">CodonU.file_handler</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../phylogenetic_analysis/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">CodonU.phylogenetic_analysis</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../../vizualizer/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">CodonU.vizualizer</span></code></a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../../index.html">CodonU</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../../index.html">API Reference</a></li>
          <li class="breadcrumb-item"><a href="../../index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">CodonU</span></code></a></li>
          <li class="breadcrumb-item"><a href="../index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">CodonU.analyzer</span></code></a></li>
      <li class="breadcrumb-item active"><code class="xref py py-mod docutils literal notranslate"><span class="pre">CodonU.analyzer.tRNA_comp</span></code></li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../../../_sources/autoapi/CodonU/analyzer/tRNA_comp/index.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="module-CodonU.analyzer.tRNA_comp">
<span id="codonu-analyzer-trna-comp"></span><h1><a class="reference internal" href="#module-CodonU.analyzer.tRNA_comp" title="CodonU.analyzer.tRNA_comp"><code class="xref py py-mod docutils literal notranslate"><span class="pre">CodonU.analyzer.tRNA_comp</span></code></a><a class="headerlink" href="#module-CodonU.analyzer.tRNA_comp" title="Link to this heading"></a></h1>
<section id="module-contents">
<h2>Module Contents<a class="headerlink" href="#module-contents" title="Link to this heading"></a></h2>
<section id="functions">
<h3>Functions<a class="headerlink" href="#functions" title="Link to this heading"></a></h3>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#CodonU.analyzer.tRNA_comp.get_anticodon_count_dict" title="CodonU.analyzer.tRNA_comp.get_anticodon_count_dict"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_anticodon_count_dict</span></code></a>(→ dict[str, int])</p></td>
<td><p>Retrieves the anticodon table from given link</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#CodonU.analyzer.tRNA_comp.calculate_gtai" title="CodonU.analyzer.tRNA_comp.calculate_gtai"><code class="xref py py-obj docutils literal notranslate"><span class="pre">calculate_gtai</span></code></a>(→ tuple[pandas.DataFrame, ...)</p></td>
<td><p>Calculates the gtAI value for each gene according to Anwar et al., 2023</p></td>
</tr>
</tbody>
</table>
<dl class="py function">
<dt class="sig sig-object py" id="CodonU.analyzer.tRNA_comp.get_anticodon_count_dict">
<span class="sig-prename descclassname"><span class="pre">CodonU.analyzer.tRNA_comp.</span></span><span class="sig-name descname"><span class="pre">get_anticodon_count_dict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">url</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">database</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">int</span><span class="p"><span class="pre">]</span></span></span></span><a class="headerlink" href="#CodonU.analyzer.tRNA_comp.get_anticodon_count_dict" title="Link to this definition"></a></dt>
<dd><p>Retrieves the anticodon table from given link</p>
<dl class="simple">
<dt><strong>NOTE:</strong> The <em>database</em> can have only two values, i.e. “<em>tRNADB_CE</em>” and “<em>GtRNAdb</em>”</dt><dd><ul class="simple">
<li><p>For using <em>tRNADB_CE</em>, please visit <a class="reference external" href="http://trna.ie.niigata-u.ac.jp/cgi-bin/trnadb/index.cgi">http://trna.ie.niigata-u.ac.jp/cgi-bin/trnadb/index.cgi</a></p></li>
<li><p>For using <em>GtRNAdb</em>, please visit <a class="reference external" href="http://gtrnadb.ucsc.edu/">http://gtrnadb.ucsc.edu/</a></p></li>
</ul>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>url</strong> – URL to anticodon table</p></li>
<li><p><strong>database</strong> – Type of database from the above options</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The dictionary containing anticodon as key and count as val</p>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="../../cua_errors/index.html#CodonU.cua_errors.UnsupportedDatabase" title="CodonU.cua_errors.UnsupportedDatabase"><strong>UnsupportedDatabase</strong></a> – If <em>database</em> has other values than mentioned</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="CodonU.analyzer.tRNA_comp.calculate_gtai">
<span class="sig-prename descclassname"><span class="pre">CodonU.analyzer.tRNA_comp.</span></span><span class="sig-name descname"><span class="pre">calculate_gtai</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">handle</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">anticodon_dict</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genetic_code_num</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">size_pop</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">60</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">generation_num</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_file</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file_name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'tAI_report'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">folder_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'Report'</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">pandas.DataFrame</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">pandas.DataFrame</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">pandas.DataFrame</span><span class="p"><span class="pre">]</span></span></span></span><a class="headerlink" href="#CodonU.analyzer.tRNA_comp.calculate_gtai" title="Link to this definition"></a></dt>
<dd><blockquote>
<div><p>Calculates the gtAI value for each gene according to Anwar et al., 2023</p>
<dl class="simple">
<dt>The function returns following dataframes:</dt><dd><ul class="simple">
<li><p>tai_df: The dataframe contains gene description and tAI values</p></li>
<li><p>abs_wi_df: The dataframe contains each anticodon and absolute weights according to the paper</p></li>
<li><p>rel_wi_df: The dataframe contains each anticodon and relative weights according to the paper</p></li>
</ul>
</dd>
</dl>
</div></blockquote>
<p><strong>Note:</strong> The function will generate a file named ‘best_fit.py’</p>
<blockquote>
<div><dl class="field-list simple">
<dt class="field-odd">param handle<span class="colon">:</span></dt>
<dd class="field-odd"><p>Path to the fasta file as a string</p>
</dd>
<dt class="field-even">param anticodon_dict<span class="colon">:</span></dt>
<dd class="field-even"><p>The dictionary containing anticodon as key and count as value</p>
</dd>
<dt class="field-odd">param genetic_code_num<span class="colon">:</span></dt>
<dd class="field-odd"><p>Genetic table number for codon table</p>
</dd>
<dt class="field-even">param reference<span class="colon">:</span></dt>
<dd class="field-even"><p>Path to the reference fasta file as a string (Optional)</p>
</dd>
<dt class="field-odd">param size_pop<span class="colon">:</span></dt>
<dd class="field-odd"><p>A parameter for the genetic algorithm to identify the population size (Optional)</p>
</dd>
<dt class="field-even">param generation_num<span class="colon">:</span></dt>
<dd class="field-even"><p>A parameter for the genetic algorithm to identify the generation number (Optional)</p>
</dd>
<dt class="field-odd">param save_file<span class="colon">:</span></dt>
<dd class="field-odd"><p>Option for saving the values in xlsx format (Optional)</p>
</dd>
<dt class="field-even">param file_name<span class="colon">:</span></dt>
<dd class="field-even"><p>Intended file name (Optional)</p>
</dd>
<dt class="field-odd">param folder_path<span class="colon">:</span></dt>
<dd class="field-odd"><p>Folder path where image should be saved (optional)</p>
</dd>
<dt class="field-even">return<span class="colon">:</span></dt>
<dd class="field-even"><p>A tuple of 3 dataframes, as discussed earlier</p>
</dd>
<dt class="field-odd">raises FileExistsError<span class="colon">:</span></dt>
<dd class="field-odd"><p>If re-write permission is not given for the file <cite>best_fit.py</cite></p>
</dd>
<dt class="field-even">raises ImportError<span class="colon">:</span></dt>
<dd class="field-even"><p>If <cite>best_fit.py</cite> is not created or deleted after creation</p>
</dd>
</dl>
</div></blockquote>
</dd></dl>

</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="../rscu_comp/index.html" class="btn btn-neutral float-left" title="CodonU.analyzer.rscu_comp" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="../../correspondence_analysis/index.html" class="btn btn-neutral float-right" title="CodonU.correspondence_analysis" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Souradipto Choudhuri.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>